src/kvstore/wal/FileBasedWal.cpp

```diff
@@ -377,6 +392,80 @@ TermID FileBasedWal::readTermId(const char* path, LogID logId) {
 }
 
 
+void FileBasedWal::scanLastWal(WalFileInfoPtr info, LogID firstId, LogID expectedLastWalId) {
+    auto* path = info->path();
+    int32_t fd = open(path, O_RDONLY);
+    if (fd < 0) {
+        LOG(FATAL) << "Failed to open file \"" << path
+                   << "\" (errno: " << errno << "): "
+                   << strerror(errno);
+    }
+
+    LogID curLogId = firstId;
+    size_t pos = 0;
+    LogID id = 0;
+    TermID term = 0;
+    int32_t head = 0;
+    int32_t foot = 0;
+    while (true) {
+        // Read the log Id
+        if (pread(fd, &id, sizeof(LogID), pos) != sizeof(LogID)) {
+            break;
+        }
+
+        if (id != curLogId) {
+            LOG(ERROR) << "LogId is not consistent" << id << " " << curLogId;
+            break;
+        }
+
+        // Read the term Id
+        if (pread(fd, &term, sizeof(TermID), pos + sizeof(LogID)) != sizeof(TermID)) {
+            break;
+        }
+
+        // Read the message length
+        if (pread(fd, &head, sizeof(int32_t), pos + sizeof(LogID) + sizeof(TermID))
+                != sizeof(int32_t)) {
+            break;
+        }
+
+        if (pread(fd, &foot, sizeof(int32_t),
+                  pos + sizeof(LogID) + sizeof(TermID) + sizeof(int32_t) + sizeof(ClusterID) + head)
+                != sizeof(int32_t)) {
+            break;
+        }
+
+        if (head != foot) {
+            LOG(ERROR) << "Message size doen't match: " << head << " != " << foot;
+            break;
+        }
+
+        info->setLastTerm(term);
+        info->setLastId(id);
+
+        if (expectedLastWalId > 0 && curLogId == expectedLastWalId) {
+            break;
+        }
+
+        // Move to the next log
+        pos += sizeof(LogID)
+               + sizeof(TermID)
+               + sizeof(ClusterID)
+               + sizeof(int32_t)
+               + head
+               + sizeof(int32_t);
+        ++curLogId;
+    }
+
+    if (pos < 0 && pos < FileUtils::fileSize(path)) {
+        LOG(WARNING) << "Invalid wal " << path << ", truncate from offset " << pos;
+        ftruncate(fd, pos);
```

Comment:
"ftruncate" only be used on file opened in write mode.  And you should check the return value for it.

Reply:
Good catch.